message/pool/message_test.go

```diff
@@ -140,3 +141,122 @@ func TestMessageMustSetPath(t *testing.T) {
 		})
 	}
 }
+
+func TestMessageAddQuery(t *testing.T) {
+	type args struct {
+		queries []string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name:    "Empty query",
+			wantErr: true,
+		},
+		{
+			name: "Single query",
+			args: args{
+				queries: []string{"a"},
+			},
+		},
+		{
+			name: "Multiple queries",
+			args: args{
+				queries: []string{"ab", "cdef", "ghijklmn"},
+			},
+		},
+		{
+			name: "Long query",
+			args: args{
+				queries: []string{strings.Repeat("q", 4096)},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			msg := pool.NewMessage()
+			for _, q := range tt.args.queries {
+				msg.AddQuery(q)
+			}
+			queries, err := msg.Queries()
+			if tt.wantErr {
+				require.Error(t, err)
+				return
+			}
+			require.NoError(t, err)
+			require.Equal(t, tt.args.queries, queries)
+		})
+	}
+}
+
+func TestMessageETag(t *testing.T) {
+	msg := pool.NewMessage()
+	require.False(t, msg.HasOption(message.ETag))
+	_, err := msg.GetETag()
+	require.Error(t, err)
+
+	etag := []byte{13, 37}
+	msg.SetETag(etag)
+	value, err := msg.GetETag()
+	require.NoError(t, err)
+	require.Equal(t, etag, value)
+
+	msg.Remove(message.ETag)
+	require.False(t, msg.HasOption(message.ETag))
+	_, err = msg.GetETag()
+	require.Error(t, err)
+
+	maxETagPathLen := int(message.CoapOptionDefs[message.ETag].MaxLen)
+	etag = make([]byte, 0)
+	for i := 1; i <= maxETagPathLen; i++ {
+		etag = append(etag, byte(i))
+		msg.SetETag(etag)
+	}
+	value, err = msg.GetETag()
+	require.NoError(t, err)
+	require.Equal(t, etag, value)
+}
+
+func TestMessageSetETag(t *testing.T) {
+	maxETagPathLen := int(message.CoapOptionDefs[message.ETag].MaxLen)
+
+	type args struct {
+		value []byte
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "Empty ETag",
+		},
+		{
+			name: "Basic ETag",
+			args: args{
+				value: []byte{13, 37},
+			},
+		},
+		{
+			name: "Long ETag",
+			args: args{
+				value: []byte(strings.Repeat("a", maxETagPathLen+1)),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			msg := pool.NewMessage()
+			msg.SetETag(tt.args.value)
+
+			value, err := msg.ETag()
```

Comment:
@jkralik we have both `GetETag` and `ETag` functions, I think that's by mistake right?

Also how strict should we take the option definition? It defines min length as 1 and max length as 8, but the setter doesn't have problem with values outside of this range.

Reply:
pls remove `GetETag`

I think we can be strict.

Reply:
`GetETag` removed.

I've added bounds check to `SetETag`, though technically that changes public API. But IMO adding `error` return value shouldn't be problematic.